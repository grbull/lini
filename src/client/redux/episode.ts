/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { EpisodeDto } from '../../server/episode/episode.dto';
import { api } from '../utils/api';

type EpisodeState = {
  isLoading: boolean;
  error?: string;
  data?: EpisodeDto;
};

const initialState: EpisodeState = {
  isLoading: false,
};

export const get = createAsyncThunk<EpisodeDto, number>('episode/get', (id) =>
  api.episode.get(id)
);

export const episode = createSlice({
  name: 'episode',
  initialState,
  reducers: {
    clear: (state) => {
      state.data = undefined;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(get.pending, (state) => {
      state.isLoading = true;
    });
    builder.addCase(get.fulfilled, (state, { payload }) => {
      state.isLoading = false;
      state.data = payload;
    });
    builder.addCase(get.rejected, (state, { error }) => {
      state.isLoading = false;
      state.error = error.message;
    });
  },
});

export const episodeActions = { ...episode.actions, get };
