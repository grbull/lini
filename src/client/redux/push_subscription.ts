/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { PushSubscriptionCreateDto } from '../../server/push_subscription/push_subscription.dto';
import { api } from '../utils/api';

type PushSubscriptionState = {
  status: 'loading' | 'idle' | 'error';
  error?: string;
};

const initialState: PushSubscriptionState = {
  status: 'idle',
};

export const create = createAsyncThunk<void, PushSubscriptionCreateDto>(
  'push_subscription/get',
  (createDto) => api.push.create(createDto)
);

export const pushSubscription = createSlice({
  name: 'push_subscription',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(create.pending, (state) => {
      state.status = 'loading';
      state.error = undefined;
    });
    builder.addCase(create.fulfilled, (state) => {
      state.status = 'idle';
    });
    builder.addCase(create.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });
  },
});

export const pushSubscriptionActions = {
  ...pushSubscription.actions,
  create,
};
