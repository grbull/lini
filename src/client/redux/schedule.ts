/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import {
  EpisodeDto,
  EpisodeScheduleDto,
} from '../../server/episode/episode.dto';
import { api } from '../utils/api';

type ScheduleState = {
  status: 'init' | 'loading' | 'idle' | 'error';
  data: {
    future: EpisodeDto[];
    past: EpisodeDto[];
  };
  error?: string;
};

const initialState: ScheduleState = {
  status: 'init',
  data: {
    future: [],
    past: [],
  },
};

export const get = createAsyncThunk<EpisodeScheduleDto>('schedule/get', () =>
  api.episode.getSchedule()
);

export const schedule = createSlice({
  name: 'schedule',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(get.pending, (state) => {
      state.status = 'loading';
      state.error = undefined;
    });
    builder.addCase(get.fulfilled, (state, { payload }) => {
      state.status = 'idle';
      state.data = payload;
    });
    builder.addCase(get.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });
  },
});

export const scheduleActions = {
  ...schedule.actions,
  get,
};
