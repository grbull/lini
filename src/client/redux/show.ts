/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { ShowWithEpisodesDto } from '../../server/show/show.dto';
import { api } from '../utils/api';

type ShowState = {
  status: 'init' | 'loading' | 'idle' | 'error';
  data?: ShowWithEpisodesDto;
  error?: string;
};

const initialState: ShowState = {
  status: 'init',
};

export const get = createAsyncThunk<ShowWithEpisodesDto, number>(
  'show/get',
  (id) => api.show.get(id)
);

export const show = createSlice({
  name: 'show',
  initialState,
  reducers: {
    reset: () => {
      return initialState;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(get.pending, (state) => {
      state.status = 'loading';
      state.data = undefined;
      state.error = undefined;
    });
    builder.addCase(get.fulfilled, (state, { payload }) => {
      state.status = 'idle';
      state.data = payload;
    });
    builder.addCase(get.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });
  },
});

export const showActions = {
  ...show.actions,
  get,
};
