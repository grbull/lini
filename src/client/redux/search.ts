/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { ShowDto } from '../../server/show/show.dto';
import { api } from '../utils/api';

type SearchState = {
  isLoading: boolean;
  error?: string;
  data?: ShowDto[];
};

const initialState: SearchState = {
  isLoading: false,
};

export const query = createAsyncThunk<ShowDto[], string>(
  'search/query',
  (query) => api.show.search(query)
);

export const search = createSlice({
  name: 'search',
  initialState,
  reducers: {
    clear: (state) => {
      state.data = undefined;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(query.pending, (state) => {
      state.isLoading = true;
    });
    builder.addCase(query.fulfilled, (state, { payload }) => {
      state.isLoading = false;
      state.data = payload;
    });
    builder.addCase(query.rejected, (state, { error }) => {
      state.isLoading = false;
      state.error = error.message;
    });
  },
});

export const searchActions = { ...search.actions, query };
