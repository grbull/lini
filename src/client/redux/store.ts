import {
  combineReducers,
  configureStore,
  getDefaultMiddleware,
} from '@reduxjs/toolkit';
import {
  FLUSH,
  PAUSE,
  PERSIST,
  persistReducer,
  persistStore,
  PURGE,
  REGISTER,
  REHYDRATE,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import { episode } from './episode';
import { pushSubscription } from './push_subscription';
import { schedule } from './schedule';
import { search } from './search';
import { show } from './show';
import { subscription } from './subscription';
import { user } from './user';

export const reducers = combineReducers({
  user: user.reducer,
  search: search.reducer,
  subscription: subscription.reducer,
  show: show.reducer,
  episode: episode.reducer,
  schedule: schedule.reducer,
  pushSubscription: pushSubscription.reducer,
});

const persistConfig = {
  key: 'root',
  storage,
  version: 2,
};

const persistedReducer = persistReducer(persistConfig, reducers);

export const store = configureStore({
  reducer: persistedReducer,
  // https://github.com/rt2zz/redux-persist/issues/988#issuecomment-552242978
  middleware: getDefaultMiddleware({
    serializableCheck: {
      ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
    },
  }),
  devTools: process.env.NODE_ENV === 'development',
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
