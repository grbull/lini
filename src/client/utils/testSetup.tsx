import { configureStore } from '@reduxjs/toolkit';
import { render, RenderResult } from '@testing-library/react';
import { createMemoryHistory, MemoryHistory } from 'history';
import React, { ReactNode } from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';

import { reducers, RootState } from '../redux/store';

export interface TestSetupProps {
  history?: MemoryHistory;
  state?: Partial<RootState>;
}

export interface TestSetupResult extends RenderResult {
  dispatch: typeof jest.fn;
  history: MemoryHistory<unknown>;
  scrollTo: typeof jest.fn;
}

export function testSetup(
  children: ReactNode,
  props?: TestSetupProps
): TestSetupResult {
  const history = props?.history || createMemoryHistory();
  const store = configureStore({
    reducer: reducers,
    preloadedState: props?.state || undefined,
  });

  // const reduxHistory = [];
  // const dispatch = (action: any): any => reduxHistory.push(action);
  const dispatch = jest.fn();
  store.dispatch = dispatch;

  const scrollTo = jest.fn();
  window.scrollTo = scrollTo;

  const utils = render(
    <Router history={history}>
      <Provider store={store}>{children}</Provider>
    </Router>
  );

  return { dispatch, history, scrollTo, ...utils };
}
