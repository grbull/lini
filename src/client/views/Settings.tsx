import cn from 'classnames';
import React, { ReactElement } from 'react';
import { useSelector } from 'react-redux';

import packageJson from '../../../package.json';
import { PageHeader } from '../components/PageHeader';
import { Preferences } from '../components/Preferences';
import { SettingsCard } from '../components/SettingsCard';
import { SettingsCardRow } from '../components/SettingsCardRow';
import { RootState } from '../redux/store';

export function Settings(): ReactElement {
  const { profile } = useSelector((state: RootState) => state.user);

  return (
    <>
      <PageHeader title="Settings" />

      <SettingsCard title="Account Information">
        <SettingsCardRow>
          <span>Email:</span>
          <span>{profile?.email}</span>
        </SettingsCardRow>

        <SettingsCardRow>
          <span>Date Registered:</span>
          <span>{profile?.dateCreated}</span>
        </SettingsCardRow>
      </SettingsCard>
      <SettingsCard title="Preferences">
        <Preferences />
      </SettingsCard>
      <SettingsCard title="Other">
        <a href={process.env.SERVER_URL + 'auth/logout'}>Logout</a>
        <div className={cn('mt-4', 'font-light', 'text-sm')}>
          Version: {packageJson.version}
        </div>
      </SettingsCard>
    </>
  );
}
