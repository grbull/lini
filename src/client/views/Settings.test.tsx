/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom/extend-expect';

import React from 'react';

import { RootState } from '../redux/store';
import { testSetup } from '../utils/testSetup';
import { Settings } from './Settings';

describe('Settings View', () => {
  it('matches the snapshot', () => {
    const { asFragment } = testSetup(<Settings />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('matches the snapshot with user data', () => {
    const initialState: Partial<RootState> = {
      user: {
        status: 'idle',
        isLoggedIn: true,
        data: {
          email: 'test@email.com',
          theme: 'auto',
          notifications: true,
          dateCreated: '2021-07-08T13:59:33.937Z',
          dateModified: '2021-07-08T13:59:33.937Z',
        },
      },
    };

    const { asFragment } = testSetup(<Settings />, { state: initialState });

    expect(asFragment()).toMatchSnapshot();
  });
});
