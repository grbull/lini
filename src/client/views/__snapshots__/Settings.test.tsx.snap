// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings View matches the snapshot 1`] = `
<DocumentFragment>
  <header
    class="w-full h-12 flex items-center p-2 md:hidden justify-between"
  >
    <button
      class="w-11 float-left"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-arrow-left fa-w-14 m-1 text-lg"
        data-icon="arrow-left"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
          fill="currentColor"
        />
      </svg>
    </button>
    <h1
      class="mx-2 font-medium text-lg overflow-x-auto whitespace-nowrap"
    >
      Settings
    </h1>
    <div
      class="w-11"
    />
  </header>
  <div
    class="p-2.5 py-3"
  >
    <h2
      class="pb-2 text-lg font-bold"
    >
      Account Information
    </h2>
    <div
      class=""
    >
      <div
        class="flex justify-between items-center py-1"
      >
        <span>
          Email:
        </span>
        <span />
      </div>
      <div
        class="flex justify-between items-center py-1"
      >
        <span>
          Date Registered:
        </span>
        <span>
          Date unavailable.
        </span>
      </div>
    </div>
  </div>
  <div
    class="p-2.5 py-3"
  >
    <h2
      class="pb-2 text-lg font-bold"
    >
      Preferences
    </h2>
    <div
      class=""
    >
      <div
        class="flex justify-between items-center py-1"
      >
        <div
          class="flex flex-col"
        >
          <span
            class="font-medium"
          >
            Theme:
          </span>
          <span
            class="font-light"
          >
            When set to auto, the app will use your system setting's to determine the theme.
          </span>
        </div>
        <span>
          <select
            aria-label="theme-select"
            class="bg-gray-500 py-1 rounded"
            id="theme"
            name="theme"
          >
            <option
              value="auto"
            >
              Auto (system)
            </option>
            <option
              value="dark"
            >
              Dark
            </option>
            <option
              value="light"
            >
              Light
            </option>
          </select>
        </span>
      </div>
      <div
        class="flex justify-between items-center py-1"
      >
        <div
          class="flex flex-col"
        >
          <span
            class="font-medium"
          >
            Notifications:
          </span>
          <span
            class="font-light"
          >
            When enabled, displays push notifications regarding soon to be aired episodes.
          </span>
        </div>
        <span>
          <div
            class="relative inline-block w-10 mr-0 align-middle select-none transition duration-200 ease-in"
          >
            <input
              aria-label="toggle"
              class="toggle-checkbox absolute block w-6 h-6 rounded-full bg-white border-4 appearance-none cursor-pointer"
              id="test"
              name="test"
              type="checkbox"
            />
            <label
              class="toggle-label block overflow-hidden h-6 rounded-full bg-gray-300 cursor-pointer"
              for="test"
            />
          </div>
        </span>
      </div>
    </div>
  </div>
  <div
    class="p-2.5 py-3"
  >
    <h2
      class="pb-2 text-lg font-bold"
    >
      Other
    </h2>
    <div
      class=""
    >
      <a
        class="block"
        download="undefineduser/data"
        href="undefineduser/data"
      >
        Download your data.
      </a>
      <a
        class="block mt-4"
        href="undefinedauth/logout"
      >
        Logout.
      </a>
      <div
        class="mt-4 font-light text-sm"
      >
        Version: 2.11.0
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Settings View matches the snapshot with user data 1`] = `
<DocumentFragment>
  <header
    class="w-full h-12 flex items-center p-2 md:hidden justify-between"
  >
    <button
      class="w-11 float-left"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-arrow-left fa-w-14 m-1 text-lg"
        data-icon="arrow-left"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
          fill="currentColor"
        />
      </svg>
    </button>
    <h1
      class="mx-2 font-medium text-lg overflow-x-auto whitespace-nowrap"
    >
      Settings
    </h1>
    <div
      class="w-11"
    />
  </header>
  <div
    class="p-2.5 py-3"
  >
    <h2
      class="pb-2 text-lg font-bold"
    >
      Account Information
    </h2>
    <div
      class=""
    >
      <div
        class="flex justify-between items-center py-1"
      >
        <span>
          Email:
        </span>
        <span>
          test@email.com
        </span>
      </div>
      <div
        class="flex justify-between items-center py-1"
      >
        <span>
          Date Registered:
        </span>
        <span>
          July 8, 2021, 2:59 PM GMT+1
        </span>
      </div>
    </div>
  </div>
  <div
    class="p-2.5 py-3"
  >
    <h2
      class="pb-2 text-lg font-bold"
    >
      Preferences
    </h2>
    <div
      class=""
    >
      <div
        class="flex justify-between items-center py-1"
      >
        <div
          class="flex flex-col"
        >
          <span
            class="font-medium"
          >
            Theme:
          </span>
          <span
            class="font-light"
          >
            When set to auto, the app will use your system setting's to determine the theme.
          </span>
        </div>
        <span>
          <select
            aria-label="theme-select"
            class="bg-gray-500 py-1 rounded"
            id="theme"
            name="theme"
          >
            <option
              selected=""
              value="auto"
            >
              Auto (system)
            </option>
            <option
              value="dark"
            >
              Dark
            </option>
            <option
              value="light"
            >
              Light
            </option>
          </select>
        </span>
      </div>
      <div
        class="flex justify-between items-center py-1"
      >
        <div
          class="flex flex-col"
        >
          <span
            class="font-medium"
          >
            Notifications:
          </span>
          <span
            class="font-light"
          >
            When enabled, displays push notifications regarding soon to be aired episodes.
          </span>
        </div>
        <span>
          <div
            class="relative inline-block w-10 mr-0 align-middle select-none transition duration-200 ease-in"
          >
            <input
              aria-label="toggle"
              class="toggle-checkbox absolute block w-6 h-6 rounded-full bg-white border-4 appearance-none cursor-pointer"
              id="test"
              name="test"
              type="checkbox"
            />
            <label
              class="toggle-label block overflow-hidden h-6 rounded-full bg-gray-300 cursor-pointer"
              for="test"
            />
          </div>
        </span>
      </div>
    </div>
  </div>
  <div
    class="p-2.5 py-3"
  >
    <h2
      class="pb-2 text-lg font-bold"
    >
      Other
    </h2>
    <div
      class=""
    >
      <a
        class="block"
        download="undefineduser/data"
        href="undefineduser/data"
      >
        Download your data.
      </a>
      <a
        class="block mt-4"
        href="undefinedauth/logout"
      >
        Logout.
      </a>
      <div
        class="mt-4 font-light text-sm"
      >
        Version: 2.11.0
      </div>
    </div>
  </div>
</DocumentFragment>
`;
